### PR TITLE
Store transfer tokens hashed in the db and only allow them to be valid for 24 hours

### PR DESCRIPTION
**Pull request by acasajus**
- Store tokens hmaced in the DB to prevent leaks even if the DB is leaked
- Only allow tokens to live for 24 hours
-

### CODE REVIEW

**Review by cquintana92 — APPROVED** [review]

File: app/dashboard/views/alias_transfer.py [hunk 1]
```diff
@@ -92,37 +104,35 @@ def alias_transfer_send_route(alias_id):
...
+    alias_transfer_url = None
 
-    # generate a new transfer_token
     if request.method == "POST":
+        # generate a new transfer_token
         if request.form.get("form-name") == "create":
-            alias.transfer_token = str(uuid4())
+            transfer_token = f"{alias.id}.{secrets.token_urlsafe(32)}"
+            alias.transfer_token = hmac_alias_transfer_token(transfer_token)
+            alias.transfer_token_expiration = arrow.utcnow().shift(minutes=30)
```
Comment (nguyenkims) [inline app/dashboard/views/alias_transfer.py:114]: I think 30 minutes is a bit too low as the other person might not be available to receive the alias immediately. 24h seems more suitable to me.

File: templates/dashboard/alias_transfer_send.html [hunk 2]
```diff
@@ -29,18 +25,40 @@ <h1 class="h3">Transfer {{ alias.email }}</h1>
           {{ alias_transfer_url }}
         </em>
 
+        <p class="mt-5">
+            Please copy the transfer URL. <strong>We won't be able to display it again</strong>. If you need to access it again you can generate a new URL.
+        </p>
+
+        <p class="mt-2">
+            This transfer URL is <strong>valid for 30 minutes</strong>. If it hasn't been used by then it will be automatically disabled.
```
Comment (nguyenkims) [inline templates/dashboard/alias_transfer_send.html:33]: If we change the 30 minutes, we need to update the wording here too.

File: templates/dashboard/alias_transfer_send.html [hunk 3]
```diff
@@ -29,18 +25,40 @@ <h1 class="h3">Transfer {{ alias.email }}</h1>
...
             If you don't want to share this alias anymore, you can remove the share URL.
           </div>
         </form>
       {% else %}
-        <form method="post">
-          <input type="hidden" name="form-name" value="create">
-          <button class="btn btn-primary mt-2">Create Share URL</button>
-        </form>
+        {% if link_active %}
+          <p class="mt-2">
```
Comment (nguyenkims) [inline templates/dashboard/alias_transfer_send.html:45]: I'd add "alert alert-info" class here so user can focus on this info.

**Review by nguyenkims — COMMENTED** [review]
LGTM, just left some comments.

File: app/dashboard/views/alias_transfer.py (hunk 1 above)
Comment (acasajus) [reply app/dashboard/views/alias_transfer.py]: Updated to 24 hours.

File: app/dashboard/views/alias_transfer.py [hunk 4]
```diff
@@ -134,12 +144,27 @@ def alias_transfer_receive_route():
...
     if not alias:
         flash("Invalid link", "error")
         return redirect(url_for("dashboard.index"))
 
+    # TODO: Don't allow none once all the tokens have been migrated to the new format
+    if (
+        alias.transfer_token_expiration is not None
+        and alias.transfer_token_expiration < arrow.utcnow()
+    ):
+        flash("Invalid link", "error")
```
[the remark this answers is not in the document]
Comment (acasajus) [reply app/dashboard/views/alias_transfer.py]: Changed to `Expired link, please request a new one`

File: templates/dashboard/alias_transfer_send.html (hunk 2 above)
Comment (acasajus) [reply templates/dashboard/alias_transfer_send.html]: changed.

File: templates/dashboard/alias_transfer_send.html (hunk 3 above)
Comment (acasajus) [reply templates/dashboard/alias_transfer_send.html]: Updated.

**Pull request merged**